core/src/test/java/org/elasticsearch/http/HttpServerTests.java

```diff
@@ -0,0 +1,254 @@
+import org.elasticsearch.common.bytes.ByteBufferBytesReference;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.BoundTransportAddress;
+import org.elasticsearch.common.transport.LocalTransportAddress;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.rest.BytesRestResponse;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Before;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+
+public class HttpServerTests extends ESTestCase {
+    private static final ByteSizeValue BREAKER_LIMIT = new ByteSizeValue(20);
+    private HttpServer httpServer;
+    private TestHttpServerTransport httpServerTransport;
+    private MemoryCircuitBreaker circuitBreaker;
+
+    @Before
+    public void setup() {
+        Settings settings = Settings.EMPTY;
+        circuitBreaker = new MemoryCircuitBreaker(BREAKER_LIMIT, 1.0d, logger);
+        httpServerTransport = new TestHttpServerTransport(circuitBreaker);
+        RestController restController = new RestController(settings);
+        restController.registerHandler(RestRequest.Method.GET, "/",
+            (request, channel) -> channel.sendResponse(new BytesRestResponse(RestStatus.OK)));
+        restController.registerHandler(RestRequest.Method.GET, "/error", (request, channel) -> {
+                throw new IllegalArgumentException("test error");
+            });
+
+        httpServer = new HttpServer(settings, httpServerTransport, restController, null) {
+            @Override
+            protected void registerHttpAddress() {
+                //no op - avoid dependency on NodeService
+            }
+
+            @Override
+            protected void unregisterHttpAddress() {
+                //no op - avoid dependency on NodeService
+            }
+        };
+        httpServer.start();
+    }
+
+    public void testDispatchRequestAddsAndFreesBytesOnSuccess() {
+        int contentLength = BREAKER_LIMIT.bytesAsInt();
+        String content = randomAsciiOfLength(contentLength);
+        TestHttpRequest request = new TestHttpRequest("/", content);
+        AssertingChannel channel = new AssertingChannel(request, true, RestStatus.OK);
+
+        httpServer.dispatchRequest(request, channel, new ThreadContext(Settings.EMPTY));
+
+        assertEquals(contentLength, httpServerTransport.maxReservedBytes());
+        assertEquals(0, httpServerTransport.reservedBytes());
+    }
+
+    public void testDispatchRequestAddsAndFreesBytesOnlyOnceOnError() {
```

Comment:
Actually I wanted to test that in case an error is thrown in a `RestHandler` (see the registered handlers in `#setup()`), we still release bytes only once when sending an error response via the `RestController`. So I think this is ok.